File: ml-class-ex1/Python-Version/ex1multi.py
#!/usr/bin/env python
# coding: utf-8

# In[2]:


###ex1multi
import numpy as np
import pandas as pd
import os
import matplotlib.pyplot as plt


# In[28]:


path = os.getcwd()+"/Documents/ML Coursera/ml-class-ex1/ex1/Python-Version/"
os.getcwd()
column_names = ['size','rooms','price']

data = pd.read_csv('ex1data2.txt', sep=",", header=None, names = column_names)


# In[78]:


data['x0'] = 1
#print(data)
X = data.iloc[:,[3,0,1]]
#print(X)
y = np.array(data.iloc[:,2])
#print(y)
theta = np.array([0,0,0])
iterations = 1500
alpha = 0.01


# In[79]:


#Cost Function already for n variables
def costFunction(X, y, theta):
    m = len(X)
    predictions = np.dot(X,theta)
    sqrErrors = np.array((predictions-y))**2
    J = 1/(2*m)*np.sum(sqrErrors)
    return J;


# In[81]:


##Normalization of variables
X.iloc[:,1] = (X.iloc[:,1]-np.mean(X.iloc[:,1]))/np.std(X.iloc[:,1])
X.iloc[:,2] = (X.iloc[:,2]-np.mean(X.iloc[:,2]))/np.std(X.iloc[:,2])
y=(y-np.mean(y))/np.std(y)


# In[104]:


def gradientDescent(X, y, theta, alpha, num_iter):
    m= len(X)
    for i in range(0, num_iter):
        theta = np.array(theta-np.transpose(alpha/m*np.dot(np.array(np.dot(X,theta)-y),X)))
    return costFunction(X, y, theta)


# In[106]:


print(gradientDescent(X,y,theta,0.01, 1500))


# In[113]:


#Normal Eq
theta=np.dot(np.dot(np.linalg.inv(np.dot(np.transpose(X),X)),np.transpose(X)),y)

File: ml-class-ex1/R version/ex1multi.R
#ex1multi

##Ex 1 ML Coursera##
library(tidyverse)
library(ggplot2)

data <- read.delim("ex1data2.txt", header = F, sep = ",")
#data%>%ggplot(aes(x = V1, y = V2))+
#  geom_point(shape = 4, color = "red")+
#  labs(x = "Population of Cityin 10,000s", y = "Profit in $10,000s")+
  
  ## Getting proper X matrix
X = cbind(data, as.vector(matrix(1,1, length(data$V1))))
X = cbind(X[,4],X[,1], X[,2])
colnames(X)<- c("x0", "size", "rooms")
head(X,2)
#Initializating rest of variables
y = as.vector(data[,3])
theta<-as.vector(rbind(0,0,0))



##Normalization features
X[,2] <- (X[,2]-mean(X[,2]))/sd(X[,2])
y <- (y-mean(y))/sd(y)
X[,3] <- (X[,3]-mean(X[,3]))/sd(X[,3])

##Vectorized Cost Function
costFunction<- function(X, y, theta){
  m <- nrow(X)
  predictions <- X%*%theta
  sqrErrors <- (predictions-y)^2
  J <- 1/(2*m)*sum(sqrErrors)
  J
}
alpha <-0.01
##Vectorized gradientDescent
theta<-as.vector(rbind(0,0,0))

gradientDescent<-function(X, y, theta, alpha, num_iter){
  m <- nrow(X)
  for(i in 1:num_iter){
    #theta - (alpha/m*sum((X*theta - y).* X))';
    theta<-as.vector(theta - t(alpha/m*colSums(as.vector(X%*%theta-y)*X)))
    costFunction(X, y, theta)
    
  }
  theta
  costFunction(X, y, theta)
  }


gradientDescent(X,y,theta,0.01, 1500)

##Normal equation
#theta = pinv((X'*X))*X'*y
#install.packages("matlib")
theta<-as.vector(solve((t(X)%*%X))%*%t(X)%*%y)
costFunction(X, y, theta)

